File: core/src/main/java/io/wwan13/domain/diaryimage/service/DiaryImageService.java
package io.wwan13.domain.diaryimage.service;

import io.wwan13.domain.diary.entity.Diary;
import io.wwan13.domain.diary.exception.DiaryNotFoundException;
import io.wwan13.domain.diary.repository.DiaryRepository;
import io.wwan13.domain.diaryimage.entity.DiaryImage;
import io.wwan13.domain.diaryimage.repository.DiaryImageRepository;
import io.wwan13.domain.image.entity.Image;
import io.wwan13.domain.image.exception.ImageNotFoundException;
import io.wwan13.domain.image.repository.ImageRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
@RequiredArgsConstructor
public class DiaryImageService {

    private final DiaryImageRepository diaryImageRepository;
    private final DiaryRepository diaryRepository;
    private final ImageRepository imageRepository;

    public Long save(Long diaryId, String imageUrl) {
        Diary diary = diaryRepository.findById(diaryId)
                .orElseThrow(() -> new DiaryNotFoundException());
        Image image = imageRepository.save(new Image(imageUrl));

        DiaryImage diaryImage = DiaryImage.builder()
                .diary(diary)
                .image(image)
                .build();
        DiaryImage createdDiaryImage = diaryImageRepository.save(diaryImage);

        return createdDiaryImage.getId();
    }
}

File: grida-core/core-domain/src/main/kotlin/org/grida/domain/diaryimage/DiaryImageService.kt
package org.grida.domain.diaryimage

import org.grida.domain.base.AccessManager
import org.grida.domain.diary.DiaryRepository
import org.grida.domain.image.Image
import org.grida.domain.image.ImageStatus
import org.grida.domain.user.UserRepository
import org.grida.error.ActivateImageAlreadyExists
import org.grida.error.GridaException
import org.grida.error.ImageGenerateAttemptOver
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class DiaryImageService(
    private val diaryImageRepository: DiaryImageRepository,
    private val diaryRepository: DiaryRepository,
    private val userRepository: UserRepository,
    private val accessManager: AccessManager,
) {

    @Transactional
    fun generateDiaryImage(
        diaryId: Long,
        userId: Long,
        imageUrl: String,
    ): Long {
        val generatedImagesCount = diaryImageRepository.countByDiaryId(diaryId)
        if (DiaryImage.canGenerateNew(generatedImagesCount)) {
            throw GridaException(ImageGenerateAttemptOver)
        }

        val diary = diaryRepository.findById(diaryId)
        accessManager.ownerOnly(diary, userId)

        val diaryImage = DiaryImage(
            userId = userId,
            diaryId = diaryId,
            image = Image(
                url = imageUrl,
                status = ImageStatus.DEACTIVATE
            )
        )
        val user = userRepository.findById(userId)
        return diaryImageRepository.save(diaryImage, diary, user)
    }

    fun countRemainImageGenerateAttempt(diaryId: Long): Long {
        val generatedImagesCount = diaryImageRepository.countByDiaryId(diaryId)
        return DiaryImage.calculateRemainingAttemptCount(generatedImagesCount)
    }

    @Transactional
    fun applyDiaryImage(
        diaryImageId: Long,
        diaryId: Long,
        userId: Long,
    ) {
        if (diaryImageRepository.existsByDiaryIdAndStatus(diaryId, ImageStatus.ACTIVATE)) {
            throw GridaException(ActivateImageAlreadyExists)
        }

        val diaryImage = diaryImageRepository.findById(diaryImageId)
        accessManager.ownerOnly(diaryImage, userId)
        diaryImageRepository.updateStatus(diaryImageId, ImageStatus.ACTIVATE)
    }

    @Transactional
    fun changeDiaryImage(
        diaryImageId: Long,
        diaryId: Long,
        userId: Long,
    ) {
        val originalDiaryImage = diaryImageRepository.findByDiaryIdAndStatus(diaryId, ImageStatus.ACTIVATE)
        diaryImageRepository.updateStatus(originalDiaryImage.id, ImageStatus.DEACTIVATE)

        val newDiaryImage = diaryImageRepository.findById(diaryImageId)
        accessManager.ownerOnly(newDiaryImage, userId)
        diaryImageRepository.updateStatus(diaryImageId, ImageStatus.ACTIVATE)
    }
}
